optee-utee/src/crypto_op.rs

```diff
@@ -0,0 +1,1048 @@
+            }
+            code => Err(Error::from_raw_error(code)),
+        }
+    }
+
+    /// both dest and tag are updated with different size
+    pub fn encrypt_final(
+        &self,
+        src: &[u8],
+        dest: &mut [u8],
+        tag: &mut [u8],
+    ) -> Result<(usize, usize)> {
+        let mut dest_size: u32 = dest.len() as u32;
+        let mut tag_size: u32 = tag.len() as u32;
+        match unsafe {
+            raw::TEE_AEEncryptFinal(
+                self.handle(),
+                src.as_ptr() as _,
+                src.len() as u32,
+                dest.as_mut_ptr() as _,
+                &mut dest_size,
+                tag.as_mut_ptr() as _,
+                &mut tag_size,
+            )
+        } {
+            raw::TEE_SUCCESS => {
+                return Ok((dest_size as usize, tag_size as usize));
+            }
+            code => Err(Error::from_raw_error(code)),
+        }
+    }
+
+    pub fn decrypt_final(&self, src: &[u8], dest: &mut [u8], tag: &[u8]) -> Result<usize> {
+        let mut dest_size: u32 = dest.len() as u32;
+        match unsafe {
+            raw::TEE_AEDecryptFinal(
+                self.handle(),
+                src.as_ptr() as _,
+                src.len() as u32,
+                dest.as_mut_ptr() as _,
+                &mut dest_size,
+                tag.as_ptr() as _,
+                tag.len() as u32,
+            )
+        } {
+            raw::TEE_SUCCESS => {
+                return Ok(dest_size as usize);
+            }
+            code => Err(Error::from_raw_error(code)),
+        }
+    }
+
+    pub fn null() -> Self {
+        Self(OperationHandle::null())
+    }
+
+    pub fn allocate(algo: AlgorithmId, mode: OperationMode, max_key_size: usize) -> Result<Self> {
+        match OperationHandle::allocate(algo, mode, max_key_size) {
+            Ok(handle) => Ok(Self(handle)),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn info(&self) -> OperationInfo {
+        self.0.info()
+    }
+
+    pub fn info_multiple(&self, info_buf: &mut [u8]) -> Result<OperationInfoMultiple> {
+        self.0.info_multiple(info_buf)
+    }
+
+    pub fn reset(&mut self) {
+        self.0.reset()
+    }
+
+    pub fn set_key<T: ObjHandle>(&self, object: &T) -> Result<()> {
+        self.0.set_key(object)
+    }
+
+    pub fn copy<T: OpHandle>(&mut self, src: &T) {
+        self.0.copy(src)
+    }
+}
+
+impl OpHandle for AE {
+    fn handle(&self) -> raw::TEE_OperationHandle {
+        self.0.handle()
+    }
+}
+
+pub struct Asymmetric(OperationHandle);
+
+impl Asymmetric {
+    /// This function can update output size with short buffer error when buffer is too
+    /// short, and example acipher utilizes this feature!
+    /// Define this function as unsafe because we need to return Ok for short buffer error.
+    pub unsafe fn encrypt(
+        &self,
+        params: &[Attribute],
+        src: &[u8],
+        dest: &mut [u8],
+    ) -> Result<usize> {
```

Comment:
`pub fn encrypt(&self,  params: &[Attribute],  src: &[u8]) -> Result<[u8]>`

Reply:
Result cannot be [u8] because of the error "error[E0277]: the size for values of type `[u8]` cannot be known at compilation time". Update the result as vector now.